### PR TITLE
HPCC-12805 JavaScript Graph Control

### PR DESCRIPTION
**Pull request by GordonSmith**
Include HPCC Visualization framework.
Add JavaScript Graph Control.
Fallback to JavaScript Graph Control when no ActiveX/NPAPI plugin exists.

Fixes HPCC-12805

Signed-off-by: Gordon Smith gordon.smith@lexisnexis.com

### CODE REVIEW

**Comment by GordonSmith** [comment]
@miguelvazq, @richardkchapman  Please Review

**Comment by richardkchapman** [comment]
Looks reasonable to me, but I'll let @miguelvazq look at it more closely before I merge.

File: esp/src/eclwatch/ESPGraph.js [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+############################################################################## */
+define([
+    "dojo/_base/declare",
+    "dojo/_base/array",
+    "dojo/_base/lang",
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: lang not used

File: esp/src/eclwatch/ESPGraph.js [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+    "dojo/i18n",
+    "dojo/i18n!./nls/hpcc",
+    "dojo/_base/Deferred",
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: Deferred not used

File: esp/src/eclwatch/ESPGraph.js [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+    "dojo/promise/all",
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: All not used

File: esp/src/eclwatch/ESPGraph.js [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+    "dojo/store/Observable",
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: Observable not used

File: esp/src/eclwatch/ESPGraph.js [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+    "dojo/topic",
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: topic not used

File: esp/src/eclwatch/ESPGraph.js [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+
+    "dojox/xml/parser",
+
+    "hpcc/WsWorkunits",
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: None of the hpcc's used??

File: esp/src/eclwatch/ESPGraph.js [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+                else if (key === "id" || key === "label")
+                    attrStr += " " + key + "=\"" + this.xmlEncode(props[key]) + "\"";
+                else
+                    propsStr += "<att name=\"" + key + "\" value=\"" + this.xmlEncode(props[key]) + "\"/>";
+            }
+            return "<node" + attrStr + ">" + propsStr + "</node>";
+        },
+
+        buildEdgeString: function (edge) {
+            var attrStr = ""
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: missing ;

File: esp/src/eclwatch/ESPGraph.js [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+                      .replace(/'/g, '&apos;')
+                      .replace(/</g, '&lt;')
+                      .replace(/>/g, '&gt;')
+                      .replace(/\n/g, '&#10;')
+                      .replace(/\r/g, '&#13;')
+            ;
+        },
+
+        buildVertexString: function (vertex, isPoint) {
+            var attrStr = ""
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: missing ;

File: esp/src/eclwatch/ESPGraph.js [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+            var propsStr = "";
+            var props = vertex.getProperties();
+            for (var key in props) {
+                if (isPoint && key.indexOf("_kind") >= 0)
+                    propsStr += "<att name=\"_kind\" value=\"point\"/>";
+                else if (key === "id" || key === "label")
+                    attrStr += " " + key + "=\"" + this.xmlEncode(props[key]) + "\"";
+                else
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: for consistency do we want to add else { }?

File: esp/src/eclwatch/ESPGraph.js [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+            var attrStr = ""
+            var propsStr = "";
+            var props = edge.getProperties();
+            for (var key in props) {
+                if (key.toLowerCase() === "id" ||
+                    key.toLowerCase() === "label" ||
+                    key.toLowerCase() === "source" ||
+                    key.toLowerCase() === "target")
+                    attrStr += " " + key + "=\"" + this.xmlEncode(props[key]) + "\"";
+                else
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: for consistency do we want to add else { }?

File: esp/src/eclwatch/ESPGraph.js [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+                    propsStr += "<att name=\"" + key + "\" value=\"" + this.xmlEncode(props[key]) + "\"/>";
+                }
+                this.m_xgmml += root ? "" : "</graph></att>" + propsStr + "</node>";
+            }
+            return false;
+        },
+
+        vertexVisited: function (vertex) {
+            if (this.m_visibleVertices[vertex.__hpcc_id]) {
+                this.m_xgmml += this.buildVertexString(vertex, false);;
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: extra ;

File: esp/src/eclwatch/ESPGraph.js [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+            }
+            else if (this.m_semiVisibleVertices[vertex.__hpcc_id]) {
+                this.m_xgmml += this.buildVertexString(vertex, true);;
```
Comment (miguelvazq) [inline esp/src/eclwatch/ESPGraph.js]: extra ;

**Comment by GordonSmith** [comment]
@miguelvazq Updated as per your comments.

File: esp/src/eclwatch/JSGraphWidget.js [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+            this.graphData = new ESPGraph();
+        },
+
+        resize: function (size) {
+            this.inherited(arguments);
+            if (this.hasPlugin()) {
+                this._plugin.graphWidget
+                    .resize()
+                    .render()
+                ;
```
Comment (miguelvazq) [inline esp/src/eclwatch/JSGraphWidget.js]: solo semicolon?

File: esp/src/eclwatch/JSGraphWidget.js [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+                                label += "+" + item.maxskew + "%, -" + item.minskew + "%";
+                            }
+                            item.__widget = new Edge()
+                                .sourceVertex(item.getSource().__widget)
+                                .targetVertex(item.getTarget().__widget)
+                                .targetMarker("arrowHead")
+                                .weight(weight)
+                                .strokeDasharray(strokeDasharray)
+                                .text(label)
+                            ;
```
Comment (miguelvazq) [inline esp/src/eclwatch/JSGraphWidget.js]: solo semicolon?

File: esp/src/eclwatch/JSGraphWidget.js [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+                            switch (item._kind) {
+                                case "point":
+                                    item.__widget = new Shape()
+                                        .radius(3)
+                                    ;
+                                    break;
+                                default:
+                                    item.__widget = new TextBox()
+                                        .text(item.label)
+                                    ;
```
Comment (miguelvazq) [inline esp/src/eclwatch/JSGraphWidget.js]: solo semicolon?

File: esp/src/eclwatch/JSGraphWidget.js [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+                        }
+                        vertices.push(subgraph.__widget);
+                    }, this);
+                    arrayUtil.forEach(this.graphData.vertices, function (item, idx) {
+                        if (!item.__widget) {
+                            switch (item._kind) {
+                                case "point":
+                                    item.__widget = new Shape()
+                                        .radius(3)
+                                    ;
```
Comment (miguelvazq) [inline esp/src/eclwatch/JSGraphWidget.js]: solo semicolon?

File: esp/src/eclwatch/JSGraphWidget.js [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+                    return this.graphData.getLocalisedXGMML(selectedItems, depth, distance);
+                },
+
+                startLayout: function (layout) {
+                    var context = this;
+                    setTimeout(function (layout) {
+                        context.graphWidget
+                            .layout("Hierarchy")
+                            .render()
+                        ;
```
Comment (miguelvazq) [inline esp/src/eclwatch/JSGraphWidget.js]: solo semicolon?

File: esp/src/eclwatch/JSGraphWidget.js [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+                KeyState_Control: 2,
+                KeyState_Menu: 4,
+
+                constructor: function (domNode) {
+                    this.graphData = new ESPGraph();
+                    this.graphWidget = new Graph()
+                        .target(domNode.id)
+                        .allowDragging(false)
+                        .render()
+                    ;
```
Comment (miguelvazq) [inline esp/src/eclwatch/JSGraphWidget.js]: solo semicolon?

**Comment by miguelvazq** [comment]
Minor comments. All else looks good.

**Comment by GordonSmith** [comment]
Those "solo semicolons" are just a style of coding which is following the Visualization widgets (which in turn are following the d3 coding styles) - each one of the method returns the object itself which allows method chaining...

**Comment by miguelvazq** [comment]
Just making sure. Approved.

**Pull request merged**